### ISSUE TITLE
Under the Reference-base model, TSS and TTS are not accurately annotated.

### ISSUE THREAD

**Issue by zpliu1126** [ablab/IsoQuant#203]
Hi~ Andrey,
When we updated isoform annotation using IsoQuant based on the existing reference annotation files, we found that it could not update the TSS and TTS of the annotated isoforms. For example, AD1_HC04_D02_425340.2 already exists in the reference annotation and is the result of IsoQuant's updated annotation. However, from the perspective of RNA read alignment, the TSS and TTS of this transcript AD1_HC04_D02_425340.2 were not well updated. Based on the expression profile calculated by Stringtie, I found that the FPKM of AD1_HC04_D02_425340.2 is only 5.6, while the FPKM of AD1_HC04_D02_425340.1 is 20; this is inconsistent with the observation results of read alignment, at least not many reads were aligned at the RI site. After I manually modified the TSS and TTS of the AD1_HC04_D02_425340.2 transcript, the FPKM of AD1_HC04_D02_425340.2 is 30 times higher than that of AD1_HC04_D02_425340.1; this seems more reasonable. Similar annotation cases are quite common throughout the genome.

![)KY3 `3BU~LIO%`0`U4YYA3](https://github.com/ablab/IsoQuant/assets/39325949/098e4933-85fe-4c7b-b579-cff696937956)

Best
zpliu

**Comment by andrewprzh** [comment]
Dear @zpliu1126 

Yes, this is a known problem (previously reported here https://github.com/ablab/IsoQuant/issues/92). If a novel isoform has a known intron chain, then TSS/TES positions are taken from the annotation, which is quite inaccurate.

It turned out to be a bit more complex than I anticipated - while reporting polyA sites is rather straightforward, detecting TSS can be non-trivial. Although we are rather short on man-power (I'm basically maintaining the project alone), I'll try to get my hands on it in the nearest future as this issue bugs me a lot.

Best
Andrey

**Comment by zpliu1126** [comment]
Dear @andrewprzh 
   > If a novel isoform has a known intron chain, then TSS/TES positions are taken from the annotation, which is quite inaccurate.
  Perhaps one could choose TSS that lead to longer transcript lengths, while utilizing polyadenylation sites (polyA sites) near transcription termination sites (TTS) to provide information. 
  From my example, it is found that inaccurate TTS and TSS will bring some interference to the comparison of downstream transcript expression.

Best
zpliu

**Comment by andrewprzh** [comment]
Probably this kind of kind of strategy makes sense, yes. In any case, it requires quite some time to implement, validate and make sure algorithmic changes do not make it worse.

I'll keep you updated on this matter!

Best
Andrey

**Comment by andrewprzh** [comment]
IsoQuant 4.0 contains substantial improvement on polyA/TSS calling. Please re-open if you encounter similar issues in the new version.

**Comment by baozg** [comment]
Did `IsoQuant` 4.0 only predict the TSS if the set `--fl_data`? I mainly work with A.thaliana, the genome is quite compact. I would like to use IsoQuant to incorpt with de novo annotation for species / samples without high-quality annotation. I am using ONT cDNA mainly which comes from dorado basecalling which automatic give FL classifciation for SQK-PCS-114 (similiar with pychopper). 

<img width="1134" height="905" alt="Image" src="https://github.com/user-attachments/assets/f5f53dad-5e9f-4200-8878-83884f37aa2b" />

Here is the IsoQuant log

```
$ cat isoquant.log                                                                                                                                                                                                          [11:59:55]
Command line: /tmp/global2/zbao/conda/envs/lranno7/bin/isoquant --reference /ebio/abt6_projects/PanBrassicaceae/tmp/Col-0_test/ref/Col-CC.fa --bam work/Col-CC/alignment/lr/CF.sup.fl.sorted.bam work/Col-CC/alignment/lr/CL_DL20.sup.f
l.sorted.bam work/Col-CC/alignment/lr/CR-1.sup.fl.sorted.bam work/Col-CC/alignment/lr/CS.sup.fl.sorted.bam --genedb work/Col-CC/experiments/eviann/all/Col-CC.protein_coding.gff3 --data_type ont --mode bulk --analysis quantification
 transcript_discovery --model_construction_strategy sensitive_ont --transcript_quantification unique_splicing_consistent --report_novel_unspliced false --report_canonical all --polya_requirement auto --min_mapq 20 --sqanti_output -
-large_output read_info read_assignments corrected_bed read2transcripts -t 32 --output work/Col-CC/experiments/isoquant/all/ont --prefix Col-CC.all.ont
2026-08-05 00:47:18,892 - INFO - Running IsoQuant version 4.0.0
2026-08-05 00:47:18,892 - WARNING - Output folder already exists, some files may be overwritten.
2026-08-05 00:47:18,938 - WARNING - work/Col-CC/experiments/isoquant/all/ont/Col-CC.all.ont folder already exists, some files may be overwritten
2026-08-05 00:47:18,939 - INFO - Novel unspliced transcripts will not be reported, set --report_novel_unspliced true to discover them
2026-08-05 00:47:18,939 - INFO -  === IsoQuant pipeline started ===
2026-08-05 00:47:18,939 - INFO - Python version: 3.9.19 | packaged by conda-forge | (main, Mar 20 2024, 12:50:21)
[GCC 12.3.0]
2026-08-05 00:47:18,939 - INFO - gffutils version: 0.13
2026-08-05 00:47:18,939 - INFO - pysam version: 0.22.1
2026-08-05 00:47:18,939 - INFO - pyfaidx version: 0.9.0.3

```

**Comment by andrewprzh** [comment]
> Did IsoQuant 4.0 only predict the TSS if the set --fl_data? 

Yes. If `--fl_data` is not set, IsoQuant assumes reads can be truncated and does not predict TSS positions as it may result in false positive calls.

Best
Andrey

**Comment by baozg** [comment]
Thanks! Did you recommend to use `--fl_data` for ONT cDNA data? For example, I didn't use it for initial testing, but even that one isoform of AT1G09420 extend UTR with closeby gene (AT1G09415). From the alignment, the boundary seems still distuiguishable

**Comment by andrewprzh** [comment]
You can give it a try, especially if you use only reads marked as FL. If you use all reads, keep in mind that predictions can be somewhat unreliable and may require additional verification (e.g. with short read data or visual confirmation).
